src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs

```diff
@@ -0,0 +1,292 @@
+        private readonly List<Func<ShellScope, Task>> _deferredTasks = new List<Func<ShellScope, Task>>();
+
+        private bool _disposeShellContext = false;
+        private bool _disposed = false;
+
+        public ShellScope(ShellContext shellContext)
+        {
+            // Prevent the context from being disposed until the end of the scope
+            Interlocked.Increment(ref shellContext._refCount);
+
+            ShellContext = shellContext;
+
+            // The service provider is null if we try to create
+            // a scope on a disabled shell or already disposed.
+            if (shellContext.ServiceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(shellContext.ServiceProvider),
+                    $"Can't resolve a scope on tenant: {shellContext.Settings.Name}");
+            }
+
+            _serviceScope = shellContext.ServiceProvider.CreateScope();
+            ServiceProvider = _serviceScope.ServiceProvider;
+        }
+
+        public ShellContext ShellContext { get; }
+        public IServiceProvider ServiceProvider { get; }
+
+        /// <summary>
+        /// Retrieve the 'ShellContext' of the current shell scope.
+        /// </summary>
+        public static ShellContext Context
+        {
+            get => Current?.ShellContext;
+        }
+
+        /// <summary>
+        /// Retrieve the 'IServiceProvider' of the current shell scope.
+        /// </summary>
+        public static IServiceProvider Services
+        {
+            get => Current?.ServiceProvider;
+        }
+
+        /// <summary>
+        /// Retrieve the current shell scope from the async flow.
+        /// </summary>
+        public static ShellScope Current
+        {
+            get => _current.Value;
+        }
+
+        /// <summary>
+        /// Start holding this shell scope along the async flow.
+        /// </summary>
+        public void StartAsyncFlow() => _current.Value = this;
+
+        /// <summary>
+        /// Execute a delegate using this shell scope.
+        /// </summary>
+        public async Task UsingAsync(Func<ShellScope, Task> execute)
+        {
+            if (Current == this)
+            {
+                await execute(Current);
+                return;
+            }
+
+            using (this)
+            {
+                StartAsyncFlow();
+
+                await ActivateShellAsync();
+
+                await execute(this);
+
+                await BeforeDisposeAsync();
+            }
+        }
+
+        /// <summary>
+        /// Activate the shell, if not yet done, by calling the related tenant event handlers.
+        /// </summary>
+        public async Task ActivateShellAsync()
+        {
+            if (ShellContext.IsActivated)
+            {
+                return;
+            }
+
+            var semaphore = _semaphores.GetOrAdd(ShellContext.Settings.Name, (name) => new SemaphoreSlim(1));
```

Comment:
This needs to be a `lock` around a standard dictionary, and also a lock in the `Remove` below.

Reply:
Okay.